torch/utils/_sympy/value_ranges.py

```diff
@@ -444,21 +360,114 @@ def fn_(x, y):
 
+class ValueRangeAnalysis(SymPyValueRangeAnalysis):
+    """ Extend SymPy Ops with IR ops """
+    def __init__(self):
+        self.name = "ValueRangeAnalysis"
+        boolean_operators = (
+            "xor",
+            "logical_and",
+            "logical_or",
+            "logical_not",
+        )
+        for op in boolean_operators:
+            setattr(self, op, self.bool_handler)
+
+    @staticmethod
+    def bool_handler(*args, **kwargs):
+        # just assuming bools can have both values
+        return ValueRanges(sympy.false, sympy.true)  # type: ignore[arg-type]
+
+    @staticmethod
+    def default_handler(*args, **kwargs):
+        # many ops are unlikely to show up in optimizable indexing compute,
+        # so we dont have full coverage
+        return ValueRanges.unknown()
+
+    def load(self, name: str, index: sympy.Expr):
+        return ValueRanges.unknown()
+
+    def store(self, name, index, value, mode=None):
+        return
+
+    def reduction(self, name, dtype, src_dtype, reduction_type, index, value):
+        return ValueRanges.unknown()
+
+    def index_expr(self, index, dtype):
+        assert isinstance(index, ValueRanges)
+        return index
+
+    @staticmethod
+    def to_dtype(x, dtype: torch.dtype):
+        x = ValueRanges.wrap(x)
+        type_ = dtype_to_type(dtype)
+        if type_ is bool:
+            if x.is_singleton():
+                return ValueRanges.wrap(x.lower != 0)
+            elif 0 not in x:
+                return ValueRanges.wrap(sympy.true)
+            else:
+                return ValueRanges(sympy.false, sympy.true)
+        # If we want to do this properly, we'd need to track the type of the variables
+        return x
+
+    @staticmethod
+    def sub(a, b):
+        b = ValueRanges.wrap(b)
+        return ValueRangeAnalysis.add(a, ValueRanges(-b.upper, -b.lower))
+
+    @staticmethod
+    def square(x):
+        return ValueRanges.convex_min_zero_map(x, lambda y: y * y)
+
+    @staticmethod
+    def neg(x):
+        return ValueRanges.decreasing_map(x, operator.neg)
+
+    @staticmethod
+    def truncdiv(a, b):
+        a = ValueRanges.wrap(a)
+        b = ValueRanges.wrap(b)
+        if 0 in b or ((-sympy.oo in a or sympy.oo in a) and (-sympy.oo in b or sympy.oo in b)):
+            return ValueRanges.unknown()
+        else:
+            # Casting to integer does truncation
+            def f(a, b):
+                result = a / b
+                # This won't work for sympy.Expr, so it'll need a workaround when
+                # dealing with dynamic shapes
+                if result.is_finite:
+                    result = sympy.Integer(result)
+                return result
+            return ValueRanges.coordinatewise_monotone_map(a, b, f)
+
+    @staticmethod
+    def sqrt(x):
+        x = ValueRanges.wrap(x)
+        if x.lower < 0:
+            return ValueRanges.unknown()
+        return ValueRanges.increasing_map(x, sympy.sqrt)
```

Comment:
Which of these changed, which are the same?

Reply:
Yeah, the diff is just terrible. Let me split the "moving things around" and the "improve things" into a few PRs